interop/test_utils.go

```diff
@@ -684,98 +685,244 @@ func DoPickFirstUnary(ctx context.Context, tc testgrpc.TestServiceClient) {
 		err = fmt.Errorf("got the reply with type %d len %d; want %d, %d", t, s, testpb.PayloadType_COMPRESSABLE, soakResponseSize)
-		return
+		return SoakIterationResult{}, err
 	}
-	return
+	// Calculate latency and return result.
+	latency := time.Since(start).Milliseconds()
+	return SoakIterationResult{
+		LatencyMs: latency,
+		Status:    "OK",
+	}, nil
 }
 
-// DoSoakTest runs large unary RPCs in a loop for a configurable number of times, with configurable failure thresholds.
-// If resetChannel is false, then each RPC will be performed on tc. Otherwise, each RPC will be performed on a new
-// stub that is created with the provided server address and dial options.
-// TODO(mohanli-ml): Create SoakTestOptions as a parameter for this method.
-func DoSoakTest(ctx context.Context, tc testgrpc.TestServiceClient, serverAddr string, dopts []grpc.DialOption, resetChannel bool, soakIterations int, maxFailures int, soakRequestSize int, soakResponseSize int, perIterationMaxAcceptableLatency time.Duration, minTimeBetweenRPCs time.Duration) {
-	start := time.Now()
-	var elapsedTime float64
-	iterationsDone := 0
-	totalFailures := 0
-	hopts := stats.HistogramOptions{
-		NumBuckets:     20,
-		GrowthFactor:   1,
-		BaseBucketSize: 1,
-		MinValue:       0,
-	}
-	h := stats.NewHistogram(hopts)
-	for i := 0; i < soakIterations; i++ {
+func executeSoakTestInThread(
+	ctx context.Context,
+	soakIterationsPerThread int,
+	startNs int64,
+	minTimeBetweenRPCs time.Duration,
+	soakRequestSize int,
+	soakResponseSize int,
+	perIterationMaxAcceptableLatency time.Duration,
+	overallTimeoutSeconds int,
+	serverAddr string,
+	threadResults *ThreadResults,
+	mu *sync.Mutex,
+	sharedChannel *grpc.ClientConn,
+	threadID int,
+	MayCreateNewChannel ManagedChannel) {
+	timeoutDuration := time.Duration(overallTimeoutSeconds) * time.Second
+	currentChannel := sharedChannel
+
+	for i := 0; i < soakIterationsPerThread; i++ {
 		if ctx.Err() != nil {
-			elapsedTime = time.Since(start).Seconds()
-			break
+			return
+		}
+
+		if time.Since(time.Unix(0, startNs)) >= timeoutDuration {
+			fmt.Printf("Test exceeded overall timeout of %d seconds, stopping...\n", overallTimeoutSeconds)
+			return
 		}
 		earliestNextStart := time.After(minTimeBetweenRPCs)
-		iterationsDone++
+		// Get the channel and client from the provided channelFunc (either shared or new).
+		_, client := MayCreateNewChannel(currentChannel)
 		var p peer.Peer
-		latency, err := doOneSoakIteration(ctx, tc, resetChannel, serverAddr, soakRequestSize, soakResponseSize, dopts, []grpc.CallOption{grpc.Peer(&p)})
-		latencyMs := int64(latency / time.Millisecond)
-		h.Add(latencyMs)
+		result, err := doOneSoakIteration(
+			ctx,
+			client,
+			soakRequestSize,
+			soakResponseSize,
+			[]grpc.CallOption{grpc.Peer(&p)})
+		addrStr := "nil"
+		if p.Addr != nil {
+			addrStr = p.Addr.String()
+		} else {
+			fmt.Fprintf(os.Stderr, "No peer address available for this RPC.\n")
+		}
```

Comment:
Can you just use `%v` or `%q` for printing instead of needing `addrStr` at all?

Reply:
I tried, but it doesn't work. An error is raised.